src/cloud/audiocom/ShareAudioDialog.cpp

```diff
@@ -0,0 +1,842 @@
+            });
+      });
+}
+
+void ShareAudioDialog::HandleUploadSucceeded(
+   std::string_view finishUploadURL, std::string_view audioSlug)
+{
+   mProgressPanel.timePanel->Hide();
+   mProgressPanel.title->SetLabel(XO("Upload complete!").Translation());
+   mProgressPanel.info->Show();
+
+   if (!GetOAuthService().HasAccessToken())
+   {
+      mProgressPanel.info->SetLabel(
+         "By pressing continue, you will be taken to audio.com and given a sharable link.");
+      mProgressPanel.info->Wrap(mProgressPanel.info->GetSize().GetWidth());
+
+      mContinueAction = [this, url = std::string(finishUploadURL)]()
+      {
+         EndModal(wxID_CLOSE);
+         OpenInDefaultBrowser({ url });
+      };
+
+      mContinueButton->Show();
+   }
+   else
+   {
+      auto sharableLink = wxString::Format(
+         "https://audio.com/%s/%s", GetUserService().GetUserSlug(),
+         audacity::ToWXString(audioSlug));
+
+      mGotoButton->Show();
+      mCloseButton->Show();
+      mCancelButton->Hide();
+
+      mGotoButton->Bind(
+         wxEVT_BUTTON,
+         [this, url = sharableLink](auto)
+         {
+            EndModal(wxID_CLOSE);
+            OpenInDefaultBrowser({ url });
+         });
+
+      mProgressPanel.link->SetValue(sharableLink);
+      mProgressPanel.linkPanel->Show();
+   }
+
+   Layout();
+   Fit();
+}
+
+void ShareAudioDialog::HandleUploadFailed(std::string_view errorMessage)
+{
+   EndModal(wxID_ABORT);
+
+   BasicUI::ShowErrorDialog(
+      {}, XO("Upload error"),
+      XO("We are unable to upload this file. Please try again and make sure to link to your audio.com account before uploading."),
+      {},
+      BasicUI::ErrorDialogOptions { BasicUI::ErrorDialogType::ModalError }.Log(
+         audacity::ToWString(errorMessage)));
+}
+
+void ShareAudioDialog::HandleExportFailure()
+{
+   EndModal(wxID_ABORT);
+
+   BasicUI::ShowErrorDialog(
+      {}, XO("Export error"),
+      XO("We are unable to prepare this file for uploading."), {},
+      BasicUI::ErrorDialogOptions { BasicUI::ErrorDialogType::ModalError });
+}
+
+void ShareAudioDialog::ResetProgress()
+{
+   mStageStartTime = Clock::now();
+   mLastUIUpdateTime = mStageStartTime;
+
+   mProgressPanel.elapsedTime->SetLabel(" 00:00:00");
+   mProgressPanel.remainingTime->SetLabel(" 00:00:00");
+   mProgressPanel.progress->SetValue(0);
+
+   mLastProgressValue = 0;
+
+   BasicUI::Yield();
+}
+
+namespace
+{
+void SetTimeLabel(wxStaticText* label, std::chrono::milliseconds time)
+{
+   wxTimeSpan tsElapsed(0, 0, 0, time.count());
+
+   label->SetLabel(tsElapsed.Format(wxT(" %H:%M:%S")));
+   label->SetName(label->GetLabel());
+   label->Update();
+}
+}
+
+void ShareAudioDialog::UpdateProgress(uint64_t current, uint64_t total)
+{
+   const auto now = Clock::now();
+
+   if (current == 0)
+      return;
+
+   if (current > total)
+      current = total;
+
+   if (mLastProgressValue != current)
+   {
+      constexpr int scale = 10000;
```

Comment:
Is this the right power of 10?

Reply:
This value is unrelated to the original progress dialog. I can use virtually any number greater that 1.

10000 is probably an overkill, but it doesn't affect the result much. Rescaling is needed because wxGauge does not allow 64 bit values.